### PR TITLE
HIVE-27169: New Locked List to prevent configuration change at runtime without throwing error

### PR DESCRIPTION
**Pull request by Aggarwal-Raghav**
### What changes were proposed in this pull request?
Create a new locked list called **_hive.conf.locked.list_** which contains comma separated configuration that won't be changed during runtime. If someone try to change them at runtime then it will give **_WARN_** log on beeline itself.

### Why are the changes needed?
In organisations, admin want to enforce some configs which user shouldn't be able to change at runtime and it shouldn't affect user's existing hive scripts. Therefore, this locked list will be useful as it will not allow user to change the value of particular configs and it will also not stop the execution of hive scripts.

### Does this PR introduce _any_ user-facing change?
This config can be documented.

### How was this patch tested?
On a cluster

### CODE REVIEW

**Comment by Aggarwal-Raghav** [comment]
@pkumarsinha, can you review this.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (okumin) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:856]: I guess this variable should be listed in the restricted list. Otherwise, users can bypass this option.
https://github.com/apache/hive/blob/539da19dc48740cb9c21cea0f8bff7ba0763e31b/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L6996-L6999

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (okumin) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:6009]: I prefer early return because it could be possible to add another restriction in the future.

File: common/src/test/org/apache/hadoop/hive/conf/TestHiveConf.java
Comment (okumin) [inline common/src/test/org/apache/hadoop/hive/conf/TestHiveConf.java:183]: I'd say we should explicitly set `hive.execution.engine=mr` here because it could be possible that tez will be the default engine someday.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (Aggarwal-Raghav) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:856]: User cannot modify _hive.conf.locked.list_ because  it is not the part of whitelist (_hive.security.authorization.sqlstd.confwhitelist_) and whitelist cannot be modified at runtime. Moreover, _hive.conf.restricted.list_ checks are enforced only after the whitelist check. Here is an example:
```
0: jdbc:hive2://localhost:10000> set hive.conf.locked.list;
+-----------------------------------------------+
|                      set                      |
+-----------------------------------------------+
| hive.conf.locked.list=hive.auto.convert.join  |
+-----------------------------------------------+
1 row selected (0.022 seconds)
0: jdbc:hive2://localhost:10000> set hive.conf.locked.list=hive.execution.engine;
Error: Error while processing statement: Cannot modify hive.conf.locked.list at runtime. It is not in list of params that are allowed to be modified at runtime (state=42000,code=1)
0: jdbc:hive2://localhost:10000>
```

File: common/src/test/org/apache/hadoop/hive/conf/TestHiveConf.java
Comment (Aggarwal-Raghav) [reply common/src/test/org/apache/hadoop/hive/conf/TestHiveConf.java:183]: Sure, will update the test.
Thanks

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (Aggarwal-Raghav) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:6009]: can you elaborate more on this?

Comment (okumin) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:6009]: I guess it is more flexible if we return here.

```
if (isLockedConfig(name)) {
  return;
}
String oldValue = name != null ? get(name) : null;
...
```

Because (1) a kind of cognitive load is lower when we read the succeeding lines, (2) we can easily reorder the priority of the whitelist vs restrict list vs locked list, and (3) we can easily add another new condition like below without an additional nest.

```
if (isLockedConfig(name)) {
  return;
}
if (isProhibitedKeyAndValue(name, value)) {
  return;
}
```

I personally don't prefer the following shape.

```
if (!isLockedConfig(name)) {
  if (!isProhibitedKeyAndValue(name, value)) {
    String oldValue = name != null ? get(name) : null;
    ...
  }
}
```

Comment (Aggarwal-Raghav) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java]: Understood

**Comment by Aggarwal-Raghav** [comment]
@okumin, I have updated the PR with the suggested changes. Can you review once?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (okumin) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:856]: There might be some security holes.

```
$ beeline
...
0: jdbc:hive2://hive-hiveserver2:10000/defaul> set hive.conf.locked.list;
+-------------------------------------+
|                 set                 |
+-------------------------------------+
| hive.conf.locked.list is undefined  |
+-------------------------------------+
1 row selected (0.179 seconds)
0: jdbc:hive2://hive-hiveserver2:10000/defaul> set hive.conf.locked.list=hive.execution.engine;
No rows affected (0.011 seconds)
0: jdbc:hive2://hive-hiveserver2:10000/defaul> set hive.conf.locked.list;
+----------------------------------------------+
|                     set                      |
+----------------------------------------------+
| hive.conf.locked.list=hive.execution.engine  |
+----------------------------------------------+
1 row selected (0.013 seconds)
```

Comment (okumin) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:856]: Looks like, that's because `hive.security.authorization.sqlstd.confwhitelist` is configured only by only a specific authorizer. I think it is better to put `hive.conf.locked.list` in `HiveConf#setupRestrictList` so that users using another authorizer can also use this feature safely.

https://github.com/apache/hive/blob/rel/release-4.0.0-beta-1/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L6909-L6914

File: ql/src/java/org/apache/hadoop/hive/ql/processors/SetProcessor.java
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/processors/SetProcessor.java:261]: I verified I could see this message via Beeline.

```
0: jdbc:hive2://hive-hiveserver2:10000/defaul> set hive.conf.locked.list;
+----------------------------------------------+
|                     set                      |
+----------------------------------------------+
| hive.conf.locked.list=hive.execution.engine  |
+----------------------------------------------+
1 row selected (0.066 seconds)
0: jdbc:hive2://hive-hiveserver2:10000/defaul> set hive.execution.engine=mr;
WARN  : Cannot modify hive.execution.engine at runtime. It is in the list of locked configurations that can't be modified at runtime
```

**Comment by okumin** [comment]
@Aggarwal-Raghav Thanks. I tested this PR on my local machine and added one more request.

Actually, I am not a committer of this project. So, please be aware that we need approval from a maintainer. Looks like, @ayushtkn left one comment in the previous PR, and it could be better to give more detailed use cases.
https://github.com/apache/hive/pull/4146#pullrequestreview-1436703783

I personally think this PR could be sometimes helpful. For example, we may sometimes ask our DWH users to configure some uncommon parameters as an ad-hoc workaround when a query hits a corner case issue. We'd like to ask them to revert the workaround once we've fixed the issue, but it is often impossible for them to do so immediately. I'm not sure it is a reason valid enough, though.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (Aggarwal-Raghav) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:856]: Yes, I am also to reproduce this behaviour and have added hive.conf.locked.list to restricted list. Thanks for the thorough testing @okumin. Have update the PR.

**Comment by Aggarwal-Raghav** [comment]
> @Aggarwal-Raghav Thanks. I tested this PR on my local machine and added one more request.
> 
> Actually, I am not a committer of this project. So, please be aware that we need approval from a maintainer. Looks like, @ayushtkn left one comment in the previous PR, and it could be better to give more detailed use cases. [#4146 (review)](https://github.com/apache/hive/pull/4146#pullrequestreview-1436703783)
> 
> I personally think this PR could be sometimes helpful. For example, we may sometimes ask our DWH users to configure some uncommon parameters as an ad-hoc workaround when a query hits a corner case issue. We'd like to ask them to revert the workaround once we've fixed the issue, but it is often impossible for them to do so immediately. I'm not sure it is a reason valid enough, though.

@okumin, nicely deduced,  this is the primary reason why I introduced this.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (okumin) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:856]: Looks good.

```
0: jdbc:hive2://hive-hiveserver2:10000/defaul> set hive.conf.locked.list;
+----------------------------------------------+
|                     set                      |
+----------------------------------------------+
| hive.conf.locked.list=hive.execution.engine  |
+----------------------------------------------+
1 row selected (0.012 seconds)
0: jdbc:hive2://hive-hiveserver2:10000/defaul> set hive.conf.locked.list=hive.auto.convert.join;
Error: Error while processing statement: Cannot modify hive.conf.locked.list at runtime. It is in the list of parameters that can't be modified at runtime or is prefixed by a restricted variable (state=42000,code=1)
```

**Review by okumin — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java
Comment (Aggarwal-Raghav) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:1360]: @pkumarsinha, we can use commons lang3 as it is present in import statement also. Will update the PR and have also verified the changes in local setup. No impact on functionality.
```
0: jdbc:hive2://localhost:10000> set hive.auto.convert.join=false;
WARN  : Cannot modify hive.auto.convert.join at runtime. It is in the list of locked configurations that can't be modified at runtime
No rows affected (0.014 seconds)

hive.log

153:2023-10-11T13:01:23,772  INFO [9c81113d-bf16-4fc7-9515-dd0790dc5f46 HiveServer2-Handler-Pool: Thread-51] session.HiveSessionImpl: executing set hive.auto.convert.join=false
158:2023-10-11T13:01:23,780  WARN [9c81113d-bf16-4fc7-9515-dd0790dc5f46 HiveServer2-Handler-Pool: Thread-51] SessionState: Cannot modify hive.auto.convert.join at runtime. It is in the list of locked configurations that can't be modified at runtime
```
Thanks

**Comment by pkumarsinha** [comment]
+1 LGTM. Ignoring one test failure because of tez container going OOM which is being investigated as HIVE-27695.

**Comment by pkumarsinha** [comment]
@Aggarwal-Raghav , There is a test failure which does not look like TezCliDriver's OOM one, can you please check if this is another flaky one or a genuine failure?

**Comment by Aggarwal-Raghav** [comment]
> @Aggarwal-Raghav , There is a test failure which does not look like TezCliDriver's OOM one, can you please check if this is another flaky one or a genuine failure?

The test is passing on my local, command used: 
```
mvn test -Dtest=TestMiniLlapLocalCliDriver -Dqfile=alterColumnStatsPart.q -pl itests/qtest -Pitests

Running org.apache.hadoop.hive.cli.TestMiniLlapLocalCliDriver
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 18.634 s - in org.apache.hadoop.hive.cli.TestMiniLlapLocalCliDriver
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  32.810 s
[INFO] Finished at: 2023-10-16T11:52:44+05:30
[INFO] ------------------------------------------------------------------------
```
Will trigger the CI again.

**Pull request merged**
